File: Java/DesignPatterns/src/main/java/designPatterns/behavioral/iterator/Repository.java
package designPatterns.behavioral.iterator;

public class Repository implements Container {
  private String[] values;

  public Repository(String[] values) {
    this.values = values;
  }

  @Override
  public Iterator getIterator() {
    return new ValueIterator(values);
  }

  private class ValueIterator implements Iterator {
    private String[] values;
    private int index;

    public ValueIterator(String[] values) {
      this.values = values;
    }

    @Override
    public boolean hasNext() {
      return index < values.length;
    }

    @Override
    public Object next() {
      if (this.hasNext()) {
        return values[index++];
      }
      return null;
    }
  }
}

File: Kotlin/DesignPatterns/src/main/kotlin/behavioral/iterator/Repository.kt
package behavioral.iterator

data class Repository(private var values: Array<String>) : Container {

  override fun getIterator(): Iterator = ValueIterator(values)

  private data class ValueIterator(private var values: Array<String>) : Iterator {
    private var index: Int = 0

    override fun hasNext(): Boolean = index < values.size

    override fun next(): Any? = if(this.hasNext()) values[index++] else null
  }
}